drivers/i2c/i2c_mspm0.c

```diff
@@ -0,0 +1,827 @@
+	} else {
+		stop = DL_I2C_CONTROLLER_STOP_DISABLE;
+	}
+
+	DL_I2C_startControllerTransferAdvanced(config->base, addr, DL_I2C_CONTROLLER_DIRECTION_RX,
+				data->transfer_len, DL_I2C_CONTROLLER_START_ENABLE, stop,
+				DL_I2C_CONTROLLER_ACK_DISABLE);
+
+	/* Wait for the read to complete */
+	k_sem_take(data->device_sync_sem, K_FOREVER);
+
+	if (data->state == I2C_MSPM0_TIMEOUT) {
+		return -ETIMEDOUT;
+	}
+
+	/* If error, return error */
+	if (((DL_I2C_getControllerStatus(config->base) & DL_I2C_CONTROLLER_STATUS_ERROR) != 0x00) ||
+	    (data->state == I2C_MSPM0_ERROR)) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int i2c_mspm0_transmit(const struct device *dev, struct i2c_msg msg, uint16_t addr)
+{
+	const struct i2c_mspm0_config *config = dev->config;
+	struct i2c_mspm0_data *data = dev->data;
+	DL_I2C_CONTROLLER_STOP stop;
+
+	data->msg_buf = msg.buf;
+	data->transfer_count = 0;
+	data->transfer_len = msg.len;
+	data->state = I2C_MSPM0_IDLE;
+
+	/* Flush anything that is left in the stale FIFO */
+	DL_I2C_flushControllerTXFIFO(config->base);
+
+	/*  Fill the FIFO
+	 *  The FIFO is 8-bytes deep, and this function will return number
+	 *  of bytes written to FIFO
+	 */
+	data->transfer_count = DL_I2C_fillControllerTXFIFO(config->base, data->msg_buf, msg.len);
+
+	/* Enable TXFIFO trigger interrupt if there are more bytes to send */
+	if (data->transfer_count < data->transfer_len) {
+		DL_I2C_enableInterrupt(config->base, DL_I2C_INTERRUPT_CONTROLLER_TXFIFO_TRIGGER);
+	} else {
+		DL_I2C_disableInterrupt(config->base, DL_I2C_INTERRUPT_CONTROLLER_TXFIFO_TRIGGER);
+	}
+
+	if (msg.flags & I2C_MSG_STOP) {
+		stop = DL_I2C_CONTROLLER_STOP_ENABLE;
+	} else {
+		stop = DL_I2C_CONTROLLER_STOP_DISABLE;
+	}
+
+	data->state = I2C_MSPM0_TX_STARTED;
+	DL_I2C_startControllerTransferAdvanced(config->base, addr, DL_I2C_CONTROLLER_DIRECTION_TX,
+				data->transfer_len, DL_I2C_CONTROLLER_START_ENABLE, stop,
+				DL_I2C_CONTROLLER_ACK_ENABLE);
+
+	/* Wait for the transmit to complete */
+	k_sem_take(data->device_sync_sem, K_FOREVER);
+
+	if (data->state == I2C_MSPM0_TIMEOUT) {
+		return -ETIMEDOUT;
+	}
+
+	if (((DL_I2C_getControllerStatus(config->base) & DL_I2C_CONTROLLER_STATUS_ERROR) != 0x00) ||
+	    (data->state == I2C_MSPM0_ERROR)) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int i2c_mspm0_transfer(const struct device *dev, struct i2c_msg *msgs, uint8_t num_msgs,
+			      uint16_t addr)
+{
+	const struct i2c_mspm0_config *config = dev->config;
+	struct i2c_mspm0_data *data = dev->data;
+	uint32_t dev_config;
+	uint8_t *internal_buf = config->merge_buf;
+	uint16_t merge_buf_size = config->merge_buf_size;
+	uint16_t current_internal_buf_size = 0;
+	uint8_t *transaction_buf;
+	uint16_t transaction_len;
+	struct i2c_msg transaction_msg;
+	int ret = 0;
+
+	if (data->is_target) {
+		/* Currently target is registered, initiating transfer is not allowed */
+		return -EBUSY;
+	}
+
+	/* Transmit each message */
+	for (int i = 0; i < num_msgs; i++) {
```

Comment:
consider splitting this block in its own function, this is getting a bit hard to follow, lot of indentation too